### PR TITLE
node taints condition check

### PR DESCRIPTION
**Pull request by amit-git**
### Node taints check

As part of the relocation service, we have a way to enable task terminations based on a configurable set of node conditions. This change is adding a configurable set of node taints that could be enabled to effect immediate task terminations. It's adding the following options to configure this behavior

1.  badTaintsPattern - a regular expression identifying undesirable node taints as reported by kubernetes
2. nodeTaintTransitionTimeThresholdSeconds - it denotes (in seconds) the least amount of time the node should sustain a taint before being eligible for task terminations

### CODE REVIEW

File: titus-supplementary-component/task-relocation/src/main/java/com/netflix/titus/supplementary/relocation/connector/AgentManagementNodeDataResolver.java [hunk 1]
```diff
@@ -95,9 +99,15 @@ private Node toNode(AgentInstanceGroup serverGroup, AgentInstance instance) {
...
+
         if (k8sNode != null) {
-            isNodeConditionBad = NodePredicates.hasBadCondition(k8sNode, badConditionMatcherFactory,
+            boolean hasBadNodeCondition = NodePredicates.hasBadCondition(k8sNode, badConditionMatcherFactory,
                     relocationConfiguration.getNodeConditionTransitionTimeThresholdSeconds());
+
+            boolean hasBadTaint = NodePredicates.hasBadTaint(k8sNode, badTaintMatcherFactory,
+                    relocationConfiguration.getNodeTaintTransitionTimeThresholdSeconds());
+
+            isNodeConditionBad = hasBadNodeCondition || hasBadTaint;
```
Comment (corindwyer) [inline titus-supplementary-component/task-relocation/src/main/java/com/netflix/titus/supplementary/relocation/connector/AgentManagementNodeDataResolver.java:110]: isNodeConditionBad -> isNodeBad

File: titus-supplementary-component/task-relocation/src/main/java/com/netflix/titus/supplementary/relocation/connector/NodePredicates.java [hunk 2]
```diff
@@ -64,11 +65,39 @@ static boolean hasBadCondition(V1Node node, Function<String, Matcher> badConditi
         return false;
     }
 
+    @VisibleForTesting
+    static boolean hasBadTaint(V1Node node, Function<String, Matcher> badTaintExpression,
+                                   int nodeTaintTransitionTimeThresholdSeconds) {
+        if (node.getSpec() != null && node.getSpec().getTaints() != null) {
+            return node.getSpec().getTaints().stream()
+                    .anyMatch(v1Taint -> badTaintExpression.apply(v1Taint.getKey()).matches() &&
+                            matchesTaintValueIfAvailable(v1Taint, Boolean.toString(true)) &&
```
Comment (corindwyer) [inline titus-supplementary-component/task-relocation/src/main/java/com/netflix/titus/supplementary/relocation/connector/NodePredicates.java:74]: I prefer `Boolean.TRUE.toString()` over `Boolean.toString(true)`. I also use a static import to get rid of the `Boolean`.

File: titus-supplementary-component/task-relocation/src/main/java/com/netflix/titus/supplementary/relocation/connector/NodePredicates.java [hunk 3]
```diff
@@ -64,11 +65,39 @@ static boolean hasBadCondition(V1Node node, Function<String, Matcher> badConditi
...
+    static boolean matchesTaintValueIfAvailable(V1Taint taint, String value) {
+        if (taint.getValue() != null) {
+            return taint.getValue().equalsIgnoreCase(value);
+        }
+        return true;
+    }
+
+
     static boolean isNodeConditionTransitionedRecently(V1NodeCondition nodeCondition, int thresholdSeconds) {
         DateTime threshold = DateTime.now().minusSeconds(thresholdSeconds);
```
Comment (corindwyer) [inline titus-supplementary-component/task-relocation/src/main/java/com/netflix/titus/supplementary/relocation/connector/NodePredicates.java:89]: by using the Titus commons Clock here, you will have more flexibility over injecting in and manipulating the time.

File: titus-supplementary-component/task-relocation/src/main/java/com/netflix/titus/supplementary/relocation/connector/NodePredicates.java [hunk 4]
```diff
@@ -64,11 +65,39 @@ static boolean hasBadCondition(V1Node node, Function<String, Matcher> badConditi
...
         if (nodeCondition.getLastTransitionTime() != null) {
             return nodeCondition.getLastTransitionTime().isAfter(threshold);
         }
         return false;
     }
+
+    static boolean isTransitionedRecently(DateTime nodeTransitionTime, int thresholdSeconds) {
+        DateTime threshold = DateTime.now().minusSeconds(thresholdSeconds);
```
Comment (corindwyer) [inline titus-supplementary-component/task-relocation/src/main/java/com/netflix/titus/supplementary/relocation/connector/NodePredicates.java:97]: same as above.

**Review by tbak — APPROVED** [review]

**Review by corindwyer — APPROVED** [review]

**Pull request merged**